nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/NiFiRecordSerDe.java

```diff
@@ -227,8 +236,32 @@ private Object extractCurrentField(Record record, RecordField field, TypeInfo fi
                 val = DataTypeUtils.convertRecordFieldtoObject(record.getValue(fieldName), field.getDataType());
                 break;
             case STRUCT:
-                val = DataTypeUtils.convertRecordFieldtoObject(record.getValue(fieldName), field.getDataType());
-                break;
+                // For some reason the Hive StandardStructObjectInspector expects the object corresponding to a "struct" to be an array or List rather than a Map.
+                // Do the conversion here, calling extractCurrentField recursively to traverse any nested structs.
+                Record r = (Record) record.getValue(fieldName);
+                if (r == null) {
+                    return null;
+                }
+                try {
+                    RecordSchema recordSchema = r.getSchema();
+                    List<RecordField> recordFields = recordSchema.getFields();
+                    if (recordFields == null || recordFields.isEmpty()) {
+                        return new ArrayList<>(0);
+                    }
+                    // This List will hold the values of the entries in the Map
+                    List<Object> structList = new ArrayList<>(recordFields.size());
+                    StructTypeInfo typeInfo = (StructTypeInfo) schema.getStructFieldTypeInfo(fieldName);
+                    for (RecordField f : recordFields) {
+                        String fName = f.getFieldName();
+                        String normalizedFieldName = fName.toLowerCase();
+                        structList.add(extractCurrentField(r, f, typeInfo.getStructFieldTypeInfo(normalizedFieldName)));
+                    }
+                    return structList;
+                } catch (Exception e) {
+                    log.warn("Error [{}] parsing Record [{}].", new Object[]{e.getLocalizedMessage(), r}, e);
```

Comment:
I would generally recommend using e.toString() instead of e.getLocalizedMessage() simply because many Exceptions don't include a message text, such as IllegalArgumentException, etc., so the message in that case comes across as "Error [] parsing Record..." or "Error [null] parsing Record..." which is confusing.